{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data/'\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EBOLA_FOLDER = DATA_FOLDER + 'ebola/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_gui = glob.glob(EBOLA_FOLDER + \"guinea_data/*.csv\")\n",
    "file_lib = glob.glob(EBOLA_FOLDER + \"liberia_data/*.csv\")\n",
    "file_sie = glob.glob(EBOLA_FOLDER + \"sl_data/*.csv\")\n",
    "list_gui = []\n",
    "#Create DataFrame for Guinea\n",
    "for file in file_gui:\n",
    "    df = pd.read_csv(file,header=0,index_col=None,parse_dates=['Date'])\n",
    "    list_gui.append(df)\n",
    "ebola_gui = pd.concat(list_gui)\n",
    "\n",
    "list_lib = []\n",
    "#Create DataFrame for Liberia\n",
    "for file in file_lib:\n",
    "    df = pd.read_csv(file,header=0,index_col=None,parse_dates=['Date'])\n",
    "    list_lib.append(df)\n",
    "ebola_lib = pd.concat(list_lib)\n",
    "\n",
    "list_sie = []\n",
    "#Create DataFrame for Sierra Leone\n",
    "for file in file_sie:\n",
    "    df = pd.read_csv(file,header=0,index_col=None,parse_dates=['date'])\n",
    "    list_sie.append(df)\n",
    "ebola_sie = pd.concat(list_sie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "res = ebola_gui.where(ebola_gui['Description'] == 'New cases of confirmed')\n",
    "df = res[pd.notnull(res['Description'])]\n",
    "new_cases_confirmed = df[['Totals', 'Date']]\n",
    "new_cases_confirmed = new_cases_confirmed.rename(columns = {'Totals': 'New cases'})\n",
    "\n",
    "res = ebola_gui.where(ebola_gui['Description'] == 'New deaths registered')\n",
    "df = res[pd.notnull(res['Description'])]\n",
    "new_deaths_registered = df[['Totals', 'Date']]\n",
    "new_deaths_registered = new_deaths_registered.rename(columns = {'Totals': 'New deaths'})\n",
    "test = new_cases_confirmed.merge(new_deaths_registered, on='Date')\n",
    "test['Country'] = 'Guinea'\n",
    "test = test.rename(columns={'Date': 'Month'})\n",
    "test['Month'] = test.Month.dt.strftime(\"%B\")\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = ebola_lib.where(ebola_lib['Variable'] == 'New case/s (confirmed)')\n",
    "df = res[pd.notnull(res['Variable'])]\n",
    "new_cases_confirmed = df[['National', 'Date']]\n",
    "new_cases_confirmed = new_cases_confirmed.rename(columns = {'National': 'New cases'})\n",
    "new_cases_confirmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = ebola_lib.where(ebola_lib['Variable'] == 'New case/s (confirmed)')\n",
    "df = res[pd.notnull(res['Variable'])]\n",
    "new_cases_confirmed = df[['National', 'Date']]\n",
    "new_cases_confirmed = new_cases_confirmed.rename(columns = {'National': 'New cases'})\n",
    "\n",
    "res = ebola_lib.where(ebola_lib['Variable'] == 'Newly reported deaths')\n",
    "df = res[pd.notnull(res['Variable'])]\n",
    "new_deaths_registered = df[['National', 'Date']]\n",
    "new_deaths_registered = new_deaths_registered.rename(columns = {'National': 'New deaths'})\n",
    "test = new_cases_confirmed.merge(new_deaths_registered, on='Date')\n",
    "test['Country'] = 'Liberia'\n",
    "test = test.rename(columns={'Date': 'Month'})\n",
    "test['Month'] = test.Month.dt.strftime(\"%B\")\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ebola_lib[list(['Date','Variable','National'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We are using 'etc_new_deaths' for computing the number of deaths per day because it's the only information we have.\n",
    "# 'etc' means 'Ebola Treatment Center', so those seem to be the deaths happening in the treatment center.\n",
    "\n",
    "res = ebola_sie.where(ebola_sie['variable'] == 'new_confirmed')\n",
    "df = res[pd.notnull(res['variable'])]\n",
    "new_cases_confirmed = df[['National', 'date']]\n",
    "new_cases_confirmed = new_cases_confirmed.rename(columns = {'National': 'New cases'})\n",
    "#mean = np.ceil(np.mean(list(map(int, new_cases_confirmed['New cases'].dropna().values))))\n",
    "#new_cases_confirmed = new_cases_confirmed.fillna(mean)\n",
    "\n",
    "res = ebola_sie.where(ebola_sie['variable'] == 'etc_new_deaths')\n",
    "df = res[pd.notnull(res['variable'])]\n",
    "new_deaths_registered = df[['National', 'date']]\n",
    "new_deaths_registered = new_deaths_registered.rename(columns = {'National': 'New deaths'})\n",
    "#mean = np.ceil(np.mean(list(map(int, new_deaths_registered['New deaths'].dropna().values))))\n",
    "#new_deaths_registered = new_deaths_registered.fillna(mean)\n",
    "total = new_cases_confirmed.merge(new_deaths_registered, on='date')\n",
    "total['Country'] = 'Sierra Leone'\n",
    "total = total.rename(columns={'date': 'Month'})\n",
    "total['Month'] = total.Month.dt.strftime(\"%B\")\n",
    "total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ebola_sie[list(['date','variable','National'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explanation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we have 9 spreedsheets containing microbiome data. This data can be seen as a single table containing 2 columns, the `NAME` and the `COUNT`. Thus, we simply read those files and concatenate them into a single dataframe representing our table with 2 columns.\n",
    "\n",
    "Given that the file does not contain headers, we ignore the first row (which is not a header) and give a name to the columns with the `header=None` and `names=['NAME', 'COUNT']` parameters.\n",
    "\n",
    "We also add a new `BARCODE` column to the dataframe, which is the name corresponding to the column name in the metadata file. We give it the same name in order to have normalized dataframes that can be merged on columns with the same name.\n",
    "\n",
    "Finally, we read the metadata file into a dataframe that we merge with our initial dataframe, we replace NA values with the `unknown` value and place an index on `[BARCODE, NAME]` in order to have unique indexes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Corresponding Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "MICROBIOME_FOLDER = DATA_FOLDER + 'microbiome/'\n",
    "mid_files = glob.glob(MICROBIOME_FOLDER + \"MID*.xls\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>COUNT</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BARCODE</th>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID1</th>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halosimplex</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronobacterium</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronomonas</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanococci Methanococcales Methanocaldococcaceae Methanotorris</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanopyri Methanopyrales Methanopyraceae Methanopyrus</th>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Thermoplasmata Thermoplasmatales Thermoplasmatales_incertae_sedis Thermogymnomonas</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Acidothermaceae Acidothermus</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Beutenbergiaceae Salana</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Bogoriellaceae Bogoriella</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cellulomonadaceae Tropheryma</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Corynebacteriaceae Turicella</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cryptosporangiaceae Cryptosporangium</th>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Dermabacter</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Devriesea</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermatophilaceae Kineosphaera</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Frankineae_incertae_sedis Fodinicola</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Blastococcus</th>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Geodermatophilus</th>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Intrasporangiaceae Kribbia</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">MID9</th>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillaceae Neptuniibacter</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillales_incertae_sedis Spongiispira</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oleiphilaceae Oleiphilus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pasteurellales Pasteurellaceae Haemophilus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Alkanindiges</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Perlucidibaca</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azomonas</th>\n",
       "      <td>805</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azorhizophilus</th>\n",
       "      <td>862</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azotobacter</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Pseudomonas</th>\n",
       "      <td>3534</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Cycloclasticus</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Sulfurivirga</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichaceae Beggiatoa</th>\n",
       "      <td>25</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichales_incertae_sedis Fangia</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Alkanibacter</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Sinobacter</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Aquimonas</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Frateuria</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Ignatzschineria</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Luteibacter</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <td>11</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <td>43</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <td>16</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            COUNT  \\\n",
       "BARCODE NAME                                                        \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      7   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...      2   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...      3   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...      3   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      7   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...      1   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      1   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      2   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...      4   \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...      1   \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...     12   \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     53   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      3   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      1   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     99   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...     17   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...      2   \n",
       "...                                                           ...   \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...     10   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...      4   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    805   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    862   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...     12   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...   3534   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      4   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...     25   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      3   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...      1   \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...      3   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...      1   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...     11   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...      1   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...     43   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...     16   \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...      1   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...      1   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...     10   \n",
       "\n",
       "                                                                         GROUP  \\\n",
       "BARCODE NAME                                                                     \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  EXTRACTION CONTROL   \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...  EXTRACTION CONTROL   \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  EXTRACTION CONTROL   \n",
       "...                                                                        ...   \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...           Control 2   \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...           Control 2   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...           Control 2   \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...           Control 2   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...           Control 2   \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...           Control 2   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...           Control 2   \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...           Control 2   \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...           Control 2   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...           Control 2   \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...           Control 2   \n",
       "\n",
       "                                                             SAMPLE  \n",
       "BARCODE NAME                                                         \n",
       "MID1    Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal...  unknown  \n",
       "        Archaea \"Crenarchaeota\" Thermoprotei Thermoprot...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Halobacteria Halobacter...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Methanococci Methanococ...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Methanopyri Methanopyra...  unknown  \n",
       "        Archaea \"Euryarchaeota\" Thermoplasmata Thermopl...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "        Bacteria \"Actinobacteria\" Actinobacteria Actino...  unknown  \n",
       "...                                                             ...  \n",
       "MID9    Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria O...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria P...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria T...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Proteobacteria\" Gammaproteobacteria X...    stool  \n",
       "        Bacteria \"Synergistetes\" Synergistia Synergista...    stool  \n",
       "        Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat...    stool  \n",
       "        Bacteria \"Tenericutes\" Mollicutes Haloplasmatal...    stool  \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    stool  \n",
       "        Bacteria \"Thermodesulfobacteria\" Thermodesulfob...    stool  \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...    stool  \n",
       "        Bacteria \"Thermotogae\" Thermotogae Thermotogale...    stool  \n",
       "        Bacteria \"Verrucomicrobia\" Opitutae Opitutales ...    stool  \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...    stool  \n",
       "        Bacteria Cyanobacteria Cyanobacteria  Chloropla...    stool  \n",
       "\n",
       "[2396 rows x 3 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initial empty dataframe\n",
    "mid = pd.DataFrame()\n",
    "for idx, file in enumerate(mid_files, 1):\n",
    "    file_name = 'MID' + str(idx)\n",
    "    # We give appropriate names to colummns and replace NA values with \"unknown\"\n",
    "    tmp_mid = pd.read_excel(MICROBIOME_FOLDER+file_name+'.xls', names=['NAME', 'COUNT'], header=None)\n",
    "    # We add a column \"BARCODE\" corresponding to the name of the file being read\n",
    "    tmp_mid['BARCODE'] = file_name\n",
    "    # We concatenate the values to create a single dataframe\n",
    "    mid = pd.concat([mid, tmp_mid])\n",
    "\n",
    "# We read the metadata file and merge it to the MID dataframe where columns have the same name\n",
    "metadata = pd.read_excel(MICROBIOME_FOLDER + 'metadata.xls')\n",
    "# We create an unique index on BARCODE and TAXON and replace NA values by 'unknown'\n",
    "final_df = pd.merge(mid, metadata).fillna('unknown').set_index(['BARCODE', 'NAME'])\n",
    "final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
